Refactor: Move ClampedPercentage type to common folder

charans29 opened the pull request:
## Description:

• Moved ClampedPercentage from `fuel-gas-price-algorithm` to `crates/types/src/` for better modularity and reuse.
• Updated references in `fuel_gas_price_provider` and `genesis` adapters.
• No functional changes, just refactoring.

## Testing:

 Ran `cargo test --all-targets`, and all tests passed.
 
 ## Linked Issue:
• close #2773

Let me know if any further adjustments are needed!

AurelienFT commented on crates/fuel-core/src/service/adapters/fuel_gas_price_provider/tests.rs:
```diff
@@ -9,12 +10,13 @@ fn build_provider<A>(
     A: Send + Sync,
 {
     let algorithm = SharedGasPriceAlgo::new_with_algorithm(algorithm);
+    // let clamped_percentage = ClampedPercentage::new(percentage as u8);
```
Can you remove this dead code please ?

charans29 replied:
Hey Aurelien, thanks for the review! I’ve removed the dead code and fixed the CI. 

I also ran the tests, and everything’s passing now. Looking forward to your feedback!

AurelienFT reviewed (commented):
Thanks for you first contribution. We really appreciate, if you can remove the little dead code and I will approve it.

Others reviewers notes : I created a PR (https://github.com/FuelLabs/fuel-core/issues/2789) to get rid of all the `as u8` we have and unify the type for percentage but it's out of the scope of this PR.

AurelienFT reviewed (approved):
Thanks a lot for your contribution. There is a little conflict if you can fix it @charans29 and I will vouch for more reviews to the team :) !

rymnc commented on crates/services/gas_price_service/src/common/fuel_core_storage_adapter.rs:
```diff
@@ -124,7 +124,10 @@ where
         + 'static,
 {
-    type Transaction<'a> = StorageTransaction<&'a mut Storage> where Self: 'a;
+    type Transaction<'a>
```
i don't think we need these changes, can you revert this hunk please?

charans29 replied:
To Fix CI, ran `cargo +nightly fmt --all ` this may result these changes

rymnc replied:
note that we use a different toolchain for formatting, which might be different than the one you're using. check the .github/workflows/ci.yml file

charans29 replied:
Thanks! I’ll reformat and push the changes accordingly. The toolchain I used is the one mentioned in Contributing.md

rymnc commented on crates/services/importer/src/ports.rs:
```diff
@@ -114,7 +114,10 @@ impl<S> Transactional for S
     S: KeyValueInspect<Column = Column>,
 {
-    type Transaction<'a> = StorageTransaction<&'a S> where Self: 'a;
+    type Transaction<'a>
```
i don't think we need these changes, can you revert this hunk please?

rymnc commented on crates/storage/src/codec/postcard.rs:
```diff
@@ -22,7 +22,10 @@ impl<T> Encode<T> for Postcard
     T: ?Sized + serde::Serialize,
 {
-    type Encoder<'a> = Cow<'a, [u8]> where T: 'a;
+    type Encoder<'a>
```
i don't think we need these changes, can you revert this hunk please?

rymnc commented on crates/storage/src/codec/raw.rs:
```diff
@@ -20,7 +20,10 @@ impl<T> Encode<T> for Raw
     T: ?Sized + AsRef<[u8]>,
 {
-    type Encoder<'a> = Cow<'a, [u8]> where T: 'a;
+    type Encoder<'a>
```
i don't think we need these changes, can you revert this hunk please?

rymnc commented on crates/types/src/clamped_percentage.rs:
```diff
@@ -0,0 +1,30 @@
+/// A value that represents a value between 0 and 100. Higher values are clamped to 100
+#[derive(
```
you can use cfg_attr instead to derive these impls conditionally if the `serde` feature is enabled. we shouldn't have serde in the default feature flag :)

charans29 replied:
used cfg_attr and removed serde from default feature

rymnc commented on tests/tests/coins.rs:
```diff
@@ -165,7 +165,7 @@ mod coin {
             .unwrap();
         assert_eq!(coins_per_asset.len(), 2);
-        assert!(coins_per_asset[0].len() >= 1);
+        assert!(coins_per_asset[0].is_empty());
```
is this a functional change in your pr?

rafal-ch replied:
I guess it's the `clippy::len_zero` lint kicking in, probably due to using a newer toolchain.

And the updated assert should say:
```suggestion
        assert!(!coins_per_asset[0].is_empty());
```

charans29 replied:
when I ran `cargo clippy --all-targets`, suggested those changes to pass all checks

rymnc commented on tests/tests/coins.rs:
```diff
@@ -613,7 +613,7 @@ mod all_coins {
             .unwrap();
         assert_eq!(coins_per_asset.len(), 2);
-        assert!(coins_per_asset[0].len() >= 1);
+        assert!(coins_per_asset[0].is_empty());
```
is this a functional change in your pr?

rafal-ch replied:
Same as https://github.com/FuelLabs/fuel-core/pull/2786#discussion_r1979787265

rymnc commented on tests/tests/gas_price.rs:
```diff
@@ -161,7 +161,7 @@ fn arb_small_tx<R: Rng + rand::CryptoRng>(
 
 #[tokio::test]
-async fn latest_gas_price__if_no_mint_tx_in_previous_block_gas_price_is_zero() {
+async fn latest_gas_price_if_no_mint_tx_in_previous_block_gas_price_is_zero() {
```
we want to retain the test naming :)

charans29 replied:
retained all non__sankecase names

rymnc commented on tests/tests/gas_price.rs:
```diff
@@ -177,7 +177,7 @@ async fn latest_gas_price__if_no_mint_tx_in_previous_block_gas_price_is_zero() {
 
 #[tokio::test]
-async fn latest_gas_price__for_single_block_should_be_starting_gas_price() {
+async fn latest_gas_price_for_single_block_should_be_starting_gas_price() {
```
same as prev

rymnc reviewed (changes requested):
good to go, please remove some of the unrelated changes :)

rafal-ch commented:
@charans29 
Thanks for you contribution!
My original comment was about using the clamped percentage [in here](https://github.com/FuelLabs/fuel-core/blob/master/bin/fuel-core/src/cli/run/tx_pool.rs#L93-L95), directly as an argument.

That's why the original issue says "Use it everywhere" ;-)

Anyway, I'm ok if this PR does not touch the place, but in such case we should remove the "closes" annotation from the description because there will be a follow-up work on the issue.

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -95,7 +95,7 @@ fn infinite_loop_tx<R: Rng + rand::CryptoRng>(
     asset_id: Option<AssetId>,
 ) -> Transaction {
-    let script = vec![op::jmp(RegId::ZERO)];
+    let script = [op::jmp(RegId::ZERO)];
```
same as https://github.com/FuelLabs/fuel-core/pull/2786#discussion_r1980890396

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -199,7 +199,7 @@ async fn latest_gas_price__for_single_block_should_be_starting_gas_price() {
 
 #[tokio::test]
-async fn produce_block__raises_gas_price() {
+async fn produce_block_raises_gas_price() {
```
done

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -251,7 +251,7 @@ async fn produce_block__raises_gas_price() {
 
 #[tokio::test]
-async fn produce_block__lowers_gas_price() {
+async fn produce_block_lowers_gas_price() {
```
done

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -303,7 +303,7 @@ async fn produce_block__lowers_gas_price() {
 
 #[tokio::test]
-async fn produce_block__dont_raises_gas_price_with_default_parameters() {
+async fn produce_block_dont_raises_gas_price_with_default_parameters() {
```
done

charans29 replied:
done

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -347,7 +347,7 @@ async fn produce_block__dont_raises_gas_price_with_default_parameters() {
 
 #[tokio::test]
-async fn estimate_gas_price__is_greater_than_actual_price_at_desired_height() {
+async fn estimate_gas_price_is_greater_than_actual_price_at_desired_height() {
```
done

charans29 commented on tests/tests/gas_price.rs:
```diff
@@ -430,7 +430,7 @@ async fn latest_gas_price__if_node_restarts_gets_latest_value() {
 
 #[tokio::test]
-async fn dry_run_opt__zero_gas_price_equal_to_none_gas_price() {
+async fn dry_run_opt_zero_gas_price_equal_to_none_gas_price() {
```
done for all

charans29 commented:
@rafal-ch 

Thanks for the feedback! Initially, I looked at PR https://github.com/FuelLabs/fuel-core/pull/2733#discussion_r1973430403 where it was proposed but not merged, so I’ve added it here with `value_parser` since `ClampedPercentage` lacks `FromStr`, keeping `fuel-core-types` minimal.

For `V0AlgorithmConfig`, `gas_price_threshold_percent` seems a good fit for `ClampedPercentage` (0–100 threshold), but I left `gas_price_change_percent` as `u64` (allows >100%) to match `v0.rs`. I reverted `v0.rs` changes (all checks passed  - verified by `cargo clippy --all-targets`, `cargo test --all-targets`) due to the mainnet warning in `metadata.rs`

I updated the tag to "ref #2773" in the description since `v0.rs` and `metadata.rs` need follow-up work I guess. Thoughts on adding `ClampedPercentage` to `V0AlgorithmConfig` in this PR, or saving it for later?

AurelienFT commented:
@MitchTurner Do you have opinion on the last comment ?

@charans29 CI isn't passing and there is few conflicts, can you fix them ? Sorry for the bunch of conflicts we have, we merged a lot of things in the last day. Thanks again for the time you take to contribute to our project.

charans29 commented:
@AurelienFT

No problem—happy to contribute! Resolved conflicts, and fixed CI, Checks passed locally and formating worked  with `cargo +nightly-2023-10-29 fmt --al` - let me know if ci fails

AurelienFT commented:
@charans29 The dependencies in the Cargo.toml of the gas price service aren't sorted correctly. (they need to be sorted in alphabetical order)

rymnc commented:
> @charans29 The dependencies in the Cargo.toml of the gas price service aren't sorted correctly. (they need to be sorted in alphabetical order)

you can use cargo-sort to achieve this. just in case there's something off with the formatting, the next command will work to fix it too

`cargo sort -w`
`npm install prettier prettier-plugin-toml && npx prettier --write "**/Cargo.toml"`

charans29 commented:
@AurelienFT @rymnc 

Thanks for the feedback! Sorted dependencies with `cargo sort -w` and formatted all `Cargo.toml` files with `npx prettier --write "**/Cargo.toml"`, and kept Rust formatting with `cargo +nightly-2023-10-29 fmt --all`. 

All checks pass locally—let me know if CI needs anything else!

rymnc commented:
> @AurelienFT @rymnc 
> 
> Thanks for the feedback! Sorted dependencies with `cargo sort -w` and formatted all `Cargo.toml` files with `npx prettier --write "**/Cargo.toml"`, and kept Rust formatting with `cargo +nightly-2023-10-29 fmt --all`. 
> 
> All checks pass locally—let me know if CI needs anything else!

fantastic, I've re-ran the checks. let's try to get this merged asap

AurelienFT commented:
@charans29 It seems that the CI still has errors :/

charans29 commented:
@AurelienFT    @rymnc 

Could you please run CI now, I’ve updated the branch and verified the following locally:

`cargo clippy --all-targets`
`cargo test -p fuel-core-tests --features aws-kms -- kms`
`cargo build -p fuel-core-bin --no-default-features --features production`
`cargo clippy --all-targets --all-features -- -D warnings`
`cargo doc --all-features --workspace --no-deps`
`cargo make check --locked`
`cargo nextest run --all-features --workspace`
`cargo nextest run --workspace`

 Let me know if anything else needs adjusting. Thanks!

AurelienFT commented:
@charans29 Thanks you fixed most of it. Eveything pass except the rust fmt with the following error : https://github.com/FuelLabs/fuel-core/actions/runs/13803567033/job/38612003153?pr=2786

rafal-ch commented:
> I updated the tag to "ref https://github.com/FuelLabs/fuel-core/issues/2773" in the description since v0.rs and metadata.rs need follow-up work I guess.

Thanks!

> Thoughts on adding ClampedPercentage to V0AlgorithmConfig in this PR, or saving it for later?

I usually prefer more but smaller PRs, but I'm also ok if you prefer to do it in a "one and done" way :)

charans29 commented:
@AurelienFT @rymnc @rafal-ch 

Fixed `rustfmt` and integrated `ClampedPercentage` into `v0` components, including `V0AlgorithmConfig`. Please check CI results—thanks!

Let me know if there’s anything else.

charans29 commented:
@AurelienFT @rymnc @rafal-ch

Integrated `ClampedPercentage` into `v0` components and fixed `rustfmt`, but CI fails:
- `publish-crates-check`: `fuel-core-types` inherits `version` from workspace, but `publish-crates` expects an explicit `version`—add to `Cargo.toml`?
- `cargo-test-kms`: AWS credentials fail despite `id-token: write` in `ci.yml`. IAM role trust policy or ARN mismatch (`024848458133` vs `249945542445`) might be the issue.

Please let me know how to proceed—thanks!

AurelienFT commented:
<!-- linear:replyTo:2657870969 -->

For this two tests it's ok they are expected to fail on a contributor PR. I will review your PR again this week. Sorry for the delay we are in the process of making a new release.

MitchTurner commented:
I apologize. I just saw this PR for the first time.

@charans29 thanks for your diligent contributions!

> For V0AlgorithmConfig, gas_price_threshold_percent seems a good fit for ClampedPercentage (0–100 threshold), but I left gas_price_change_percent as u64 (allows >100%) to match v0.rs. I reverted v0.rs changes (all checks passed - verified by cargo clippy --all-targets, cargo test --all-targets) due to the mainnet warning in metadata.rs

I think your instincts here are right. I don't think everything in our codebase that is a percentage qualifies as a `ClampedPercentage`. I introduced this type to avoid certain mathematical confusions in the gas price code. TBH I don't think I'm in favor of moving it into the `fuel-types` crate for that reason. Unifying them within the gas price algorithm crate however seems like a great idea.

```rs
    /// The Percentage the execution gas price will change in a single block, either increase or decrease
    /// based on the fullness of the last L2 block
    pub exec_gas_price_change_percent: ClampedPercentage,
```
Also, there are some places that you converted (like) above that were intentionally left as  not `ClammpedPercentage`. I'd prefer to leave all places the same types.

charans29 commented:
@MitchTurner @AurelienFT @rafal-ch

Thank you, @MitchTurner, for your feedback! I moved `ClampedPercentage` from `crates/fuel-gas-price-algorithm/src/v1.rs` to `crates/types/src/clamped_percentage.rs` and applied it across the project per #2773’s goal of project-wide adoption (@AurelienFT). I understand you’d prefer it remains in `fuel-gas-price-algorithm` to avoid confusion in gas price calculations and that all instances I converted should retain their original types.

I’d appreciate your guidance:
- Should I relocate `ClampedPercentage` to `crates/fuel-gas-price-algorithm` and revert all changes made in this PR?

@AurelienFT, is #2773’s objective of project-wide use still active? @rafal-ch, you supported `v0` changes—any thoughts? I’ll update the PR based on your collective input—thank you!

rymnc commented:
hi @charans29 do you mind updating the base branch of your PR?

rafal-ch commented:
@charans29 
Sorry for radio-silence. We'll get back to you shortly after agreeing the way to go, since we don't want you to go back and forth :)

charans29 commented:
@rymnc @rafal-ch

Hi, @rymnc, I’ve rebased `charans29/use_clamped_percentage` onto `upstream/master` and

@rafal-ch, no problem on the silence, looking forward to your guidance once aligned

AurelienFT commented:
Thanks a lot @charans29 for your work on that PR !  Really appreciated
